### ISSUE TITLE
Orion in Docker Container form when running in fail state then not start by existing pid

### ISSUE THREAD

**Issue by cesarjorgemartinez** [telefonicaid/fiware-orion#3166]
Orion in Docker Container form when running in fail state then not start by existing pid

Example log:

time=Wednesday 25 Apr 08:06:39 2018.654Z | lvl=ERROR | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=contextBroker.cpp[1312]:pidFile | msg=PID-file '/tmp/contextBroker.pid' found. A broker seems to be running already

**Comment by fgalan** [comment]
Two possible solutions:

1. Modify Dockerfile so a rm for pid file is done before starting B
2. Add a way to disable pidfile generation in Orion CLI

**Comment by cdupont** [comment]
I have the same issue.
Not sure to understand though: what's happening here? What do you mean by "running in fail state"?

**Comment by cesarjorgemartinez** [comment]
Provissional workaround by @cesarjorgemartinez:
`sudo docker rm iot-orion`
Then:
`docker-compose up .....`

**Comment by fgalan** [comment]
In theory, if during `docker stop` the OS in the container (where Orion runs) sends a SIGINT to Orion process, then Orion will remove the .pid file before stopping. That should be the standard procedure.

**Comment by cesarjorgemartinez** [comment]
Reproduced in Uma environment this issue. Apply the workaroud:
sudo docker rm iot-orion
up...

**Comment by jmezzera** [comment]
Hi,
Im having the same issue here. 

> In theory, if during docker stop the OS in the container (where Orion runs) sends a SIGINT to Orion process, then Orion will remove the .pid file before stopping. That should be the standard procedure.

@fgalan, I get that is the standard procedure but what about non standard e.g. an unexpected server reboot? How could I ensure that Orion is going to start correctly after a reboot?

Thanks

**Comment by fgalan** [comment]
> @fgalan, I get that is the standard procedure but what about non standard e.g. an unexpected server reboot? How could I ensure that Orion is going to start correctly after a reboot?

An ordered reboot (i.e. "sudo reboot" in a typical Linux system) causes the OS stopping its processes in a proper way (i.e. OS sends SIGINT to all the processes). I understand you refer a more critical system fail, i.e. a hard disk crashes, electrical power outage or something like that.

If such critical even occurs, you should include clean up procedures as part as you recovery plan. In particular, the cleanup of dirty pid files should be included (including the one from CB). That should avoid any problem with CB restart.

Note that CB is checking the pid for a good reason: to avoid re-rerun the same CB instance twice.

**Comment by jmezzera** [comment]
> I understand you refer a more critical system fail, i.e. a hard disk crashes, electrical power outage or something like that.

I was indeed referring to such a failure.

> you should include clean up procedures as part as you recovery plan.

Fair enough, I get what you mean.
Thanks for your response

**Comment by cesarjorgemartinez** [comment]
Hi, when will this issue be fixed?

**Comment by fgalan** [comment]
This is not an issue that needs to be fixed :) As I have mentiened in previous comments, this is not an issue/bug, but a feature, that prevents several CBs to run at the same time due to errors or bad configuration.

There would be an issue if in the event of an ordered shutdown of CB (i.e. upong getting and SIGINT or SIGTERM signal) the PID file would not be removed by CB. Is that the situation?

Otherwise, in the case of a crash the system doing the restart (tooling scripts?) should cleanup the environment upon recovering (not only Orion PID file, other processes could also have lefovers) before starting any process. That belong to the software doing the restart and is out of scope of Orion.

**Comment by jmezzera** [comment]
@cesarjorgemartinez I understood the point @fgalan made and now I agree with him. I now have the OS in charge of cleaning a failed termination. I hope this helps.

**Comment by cesarjorgemartinez** [comment]
From the point of view of DevOps it is a big issue.

One container that failed abruptly, the same container need to restore this (pid). 

if this is not done, an unavailability is caused and it must be someone manually who restores the container, and that is unacceptable by the end customer.

Do you need me to review and correct this problem with a pull request?

Regards,
Cesar Jorge

**Comment by fgalan** [comment]
> One container that failed abruptly, the same container need to restore this (pid).

In this case, the container (not Orion) should have in its re-starting logic (entrypoint.sh?) the cleanup of the file (with rm).

**Comment by karikolehmainen** [comment]
This seems to be an issue still. Is there are way to add rm command in docker compose  file in command: -section? It happens more than it should that either virtual machine or machine (raspberry) where container is running makes hard reset and recovering from that requires manual work. There should be at least environment switch that would either clean PID file before startup or disregard PID file altogether.

**Comment by cdupont** [comment]
@karikolehmainen I do remove the PID in the docker-compose:
```
  orion:
    image: fiware/orion:2.2.0
    hostname: orion
    ports:
      - "1026:1026"
    entrypoint: ["sh", "-c", "rm /tmp/contextBroker.pid; /usr/bin/contextBroker -fg -multiservice -dbhost localhost -logLevel DEBUG"]
    restart: always
    network_mode: host
    depends_on:
      - "mongo"
 ```
See the entrypoint: `rm /tmp/contextBroker.pid;`.

@fgalan just a side note, maybe it would be nice to edit the title, I had to read it 3 times to understand it :)

**Comment by fgalan** [comment]
> @fgalan just a side note, maybe it would be nice to edit the title, I had to read it 3 times to understand it :)

You mean the tittle of the issue? Which one do you suggest to make it clearer?

**Comment by cdupont** [comment]
Something like: "Orion docker container won't start because of existing PID"

**Comment by fgalan** [comment]
> Something like: "Orion docker container won't start because of existing PID"

Changed!

**Comment by kzangeli** [comment]
FYI, I completely removed the PID-file from Orion-LD.
It's not really necessary, as if a broker is already is running, the port is occupied and restStart will fail.

[ Not true if you start two different brokers on different ports ]

**Comment by fgalan** [comment]
> This seems to be an issue still. Is there are way to add rm command in docker compose file in command: -section? It happens more than it should that either virtual machine or machine (raspberry) where container is running makes hard reset and recovering from that requires manual work. There should be at least environment switch that would either clean PID file before startup or disregard PID file altogether.

If it happens that the virtual machine or machine (raspberry) where Orion runs is hard resetting the container, then the actual problem may be in the virtual machine or machine (raspberry). In this case, the PID checking is working nicely as a safety measure: it is telling you that something is not healthy in your environment.

...and if you want to intentionally disable that safety measure (the real world of production system sometimes lead you to do dirty things :) then the modifications to Dockerfile proposed by @cdupont looks nice.

**Comment by cesarjorgemartinez** [comment]
We reproduced this bug on 2022/09/02 after reboot the environment when fixing a system full disk problem. Version of Orion is 3.7.0.

**Comment by fgalan** [comment]
This is not a bug but a desired behaviour, as I have explained [here](https://github.com/telefonicaid/fiware-orion/issues/3166#issuecomment-553295349) and [here](https://github.com/telefonicaid/fiware-orion/issues/3166#issuecomment-539434852).
